### PR TITLE
CalDAV to sync properly when limit is set in PDO backend

### PR DESCRIPTION
**Pull request by nhirokinet**
Hi,

I altered some codes to make CalDAV sync properly even when limit is set. Changes is mainly as below two:

- Return syncToken when limit is set. Current code returns the syncToken of the latest server state. Changed to return the syncToken corresponding to the truncated sync result.
  - When the result limit is designated by client, this is not the consistent as returned result.
  - In [RFC6578 section 3.6](https://tools.ietf.org/html/rfc6578#section-3.6): `the DAV:sync-token value returned in the response MUST represent the correct state for the partial set of
   changes returned.`
  - So, when truncation occurs, client will miss the truncated event even in later sync, since the client receives the syncToken of not received events, and asks server to send the event after that next time.
- When response is truncated, 507 record is added to the response.
  - Defined in [RFC6578 section 3.6](https://tools.ietf.org/html/rfc6578#section-3.6)
  - With this, the client can recognize the result is truncated, and can retry until the client syncs with the latest state.
  - For example, OneCalendar requested once I press sync button in current code. In new code OneCalendar repeated until it reaches the latest state, so I just had to press sync button once.

OneCalendar in Windows Store sends a request with limit 1000, and I checked using this, but with the changes below:

- I defanged the code of Authentication because OneCalendar did not send a request with Digest authentication.
- I overrided the limit to 2 in `lib/CalDAV/Backend/PDO.php` while testing.

I think CardDAV has the same issue, but I did not find an environment to reproduce and test with, so I wrote a code only for CalDAV, PDO backend.

This is my first code around WebDAV, so could you take a look including my understanding is correct or not?

### CODE REVIEW

**Comment by evert** [comment]
It's a bit hard for me to eye this and see if it's correct, but it *looks* good. I feel that this can use a good amount of unittests to ensure that edge cases are sufficiently covered and increase confidence.

**Comment by nhirokinet** [comment]
I added test case for each of two changes.

**Comment by nhirokinet** [comment]
@DeepDiver1975 I added some tests after your comments, should I add some more change?

**Comment by nhirokinet** [comment]
Hi, is there something I can do to advance this pull request?
CI fails, but it seems to be some new timezone test, which this pull request does not touch.

**Comment by nhirokinet** [comment]
@DeepDiver1975 Hi, do you have some comments about added tests?

**Comment by nhirokinet** [comment]
@phil-davis @DeepDiver1975 Hi, do you know the status about this? I added some tests according to your comment, long ago, that I'm not sure about my memory, anything more I should do?

File: tests/Sabre/DAV/Sync/PluginTest.php [hunk 1]
```diff
@@ -224,13 +224,18 @@ public function testSubsequentSyncSyncCollectionLimit()
         );
 
         $responses = $multiStatus->getResponses();
-        $this->assertEquals(1, count($responses), 'We expected exactly 1 {DAV:}response');
+        $this->assertEquals(2, count($responses), 'We expected exactly 1 {DAV:}response');
```
Comment (DeepDiver1975) [inline tests/Sabre/DAV/Sync/PluginTest.php:227]: > We expected exactly 1 {DAV:}response

We expected exactly 2 {DAV:}response

**Review by DeepDiver1975 — CHANGES_REQUESTED** [review]

File: tests/Sabre/DAV/Sync/PluginTest.php (hunk 1 above)
Comment (nhirokinet) [reply tests/Sabre/DAV/Sync/PluginTest.php]: fixed the message.

**Review by DeepDiver1975 — APPROVED** [review]

**Review by phil-davis — APPROVED** [review]

**Comment by phil-davis** [comment]
Rebased and squashed. LGTM, thanks. Merging.
